### PR TITLE
Show Pull Request button or status of latest PR in branch list

### LINKED ISSUE #4547: Proposal: quick compare buttons at branch list
- Gitea version (or commit ref): any
- Git version: any
- Operating system: any
- Database: any
- Can you reproduce the bug at https://try.gitea.io: no (server down)

## Description

On /{user}/{repo}/branches  page, there is brief list of branches.

Github shows compare button or pull request link for each non-default branch, which compares branch to default branch (usually master) and proposes to create Pull request (or links to pull request itself).

This functionality is missing in gitea. Would be good to implement this in gitea.

### PR DESCRIPTION
**Pull request by saitho**
Resolves #4547.

If a branch already has a pull request it will be shown instead of the "Create PR" button with its status (open, merged, closed). Similarly to GitHub, the latest pull request for a branch will be shown there.
Deleted branches will not have the pull request button.

Also, as a little follow-up on #6695, I've also hidden the commit divergence section for deleted branches as it won't contain useful information since the branch is deleted.

## Screenshot

branch_list-pull_requests-smaller_buttons

### CODE REVIEW

**Comment by saitho** [comment]
Failing due to the Git for libravatar being down. :(

**Comment by saitho** [comment]
> Failing due to the Git for libravatar being down. :(

Should work again. Can someone restart the build? :)

**Comment by lunny** [comment]
restarted

File: models/pull.go [hunk 1]
```diff
@@ -1037,6 +1037,21 @@ func GetUnmergedPullRequestsByHeadInfo(repoID int64, branch string) ([]*PullRequ
...
 
+// GetLatestPullRequestsByHeadInfo returns the latest pull request (regardless of its status)
+// by given head information (repo and branch).
+func GetLatestPullRequestsByHeadInfo(repoID int64, branch string) (*PullRequest, error) {
+	prs := make([]*PullRequest, 0, 2)
+	err := x.
+		Where("head_repo_id = ? AND head_branch = ?", repoID, branch).
+		OrderBy("id DESC").
+		Limit(1).
+		Find(&prs)
```
Comment (lunny) [inline models/pull.go]: Why not just use `Get(&pr)` and remove `Limit(1)`?

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/pull.go (hunk 1 above)
Comment (saitho) [reply models/pull.go]: True... Didn't look at the lines above where a single entry is fetched.
Just adjusted that via the GitHub editor – i.e. I didn't test if it still works (will do later).

**Comment by lunny** [comment]
Could we give a link to that PR?

File: templates/repo/branch/list.tmpl [hunk 2]
```diff
@@ -57,6 +59,27 @@
...
+									<td class="right aligned">
+
+									{{if not .LatestPullRequest}}
+										{{if not .IsDeleted}}
+										<a href="{{$.RepoLink}}/compare/{{$.DefaultBranch | EscapePound}}...{{if ne $.Repository.Owner.Name $.Owner.Name}}{{$.Owner.Name}}:{{end}}{{.Name | EscapePound}}">
+                                        	<button id="new-pull-request" class="ui compact basic button">{{$.i18n.Tr "repo.pulls.compare_changes"}}</button>
+                                        </a>
+										{{end}}
+									{{else}}
+										#{{.LatestPullRequest.Issue.Index}}
```
Comment (silverwind) [inline templates/repo/branch/list.tmpl]: Would make this a link too, like on GitHub.

File: templates/repo/branch/list.tmpl [hunk 3]
```diff
@@ -57,6 +59,27 @@
...
+                                        {{if .LatestPullRequest.HasMerged}}
+                                        	<a href="{{$.RepoLink}}/pulls/{{.LatestPullRequest.Issue.Index}}" class="ui purple large label"><i class="octicon octicon-git-pull-request"></i> {{$.i18n.Tr "repo.pulls.merged"}}</a>
```
Comment (silverwind) [inline templates/repo/branch/list.tmpl]: Can you make these buttons a tiny bit smaller (less padding)? Maybe by just removing the `large` class.

Comment (silverwind) [inline templates/repo/branch/list.tmpl]: Can you make these buttons a tiny bit smaller (less padding)? Maybe by just removing the `large` class.

**Comment by silverwind** [comment]
Also something to check: Github disabled the delete button on branches with open pull requests (probably because the PR would become invalid). Maybe this is something we should consider too.

File: templates/repo/branch/list.tmpl (hunk 3 above)
Comment (saitho) [reply templates/repo/branch/list.tmpl]: I'll look into that on ~Saturday :)

**Comment by saitho** [comment]
> Also something to check: Github disables the branch delete button on branches with open pull requests (probably because the PR would become invalid). Maybe this is something we should consider doing too.

Didn't to that for now, since the user has to confirm in the popup he REALLY wants to delete the branch. ;)

**Review by zeripath — APPROVED** [review]

File: routers/repo/branch.go [hunk 4]
```diff
@@ -178,12 +179,18 @@ func loadBranches(ctx *context.Context) []*Branch {
 			return nil
 		}
 
+		pr, err := models.GetLatestPullRequestsByHeadInfo(ctx.Repo.Repository.ID, branchName)
```
Comment (lunny) [inline routers/repo/branch.go]: err didn't be handled.

**Review by lunny — CHANGES_REQUESTED** [review]

File: templates/repo/branch/list.tmpl [hunk 5]
```diff
@@ -44,9 +45,10 @@
 									{{else}}
 										<a href="{{$.RepoLink}}/src/branch/{{.Name | EscapePound}}">{{.Name}}</a>
 										<p class="time">{{$.i18n.Tr "org.repo_updated"}} {{TimeSince .Commit.Committer.When $.i18n.Lang}}</p>
-									</td>
 									{{end}}
+									</td>
 									<td class="ui">
+										{{if not .IsDeleted}}
```
Comment (silverwind) [inline templates/repo/branch/list.tmpl:51]: Is it necessary to render this empty `<td class="ui">` when this condition is false?

File: templates/repo/branch/list.tmpl [hunk 6]
```diff
@@ -57,6 +59,27 @@
                                             	<div class="bar bar-ahead" style="width: {{percentage .CommitsAhead .CommitsBehind .CommitsAhead}}%"></div>
 											</div>
 										</div>
+										{{end}}
+									</td>
+									<td class="right aligned">
+
+									{{if not .LatestPullRequest}}
+										{{if not .IsDeleted}}
+										<a href="{{$.RepoLink}}/compare/{{$.DefaultBranch | EscapePound}}...{{if ne $.Repository.Owner.Name $.Owner.Name}}{{$.Owner.Name}}:{{end}}{{.Name | EscapePound}}">
```
Comment (silverwind) [inline templates/repo/branch/list.tmpl]: Intend here for consistency.

File: templates/repo/branch/list.tmpl (hunk 5 above)
Comment (saitho) [reply templates/repo/branch/list.tmpl:51]: Yes, as the list may contain existing and deleted branches.

**Comment by saitho** [comment]
Needs one more approval. :)

File: routers/repo/branch.go [hunk 7]
```diff
@@ -181,12 +182,22 @@ func loadBranches(ctx *context.Context) []*Branch {
 			return nil
 		}
 
+		pr, err := models.GetLatestPullRequestByHeadInfo(ctx.Repo.Repository.ID, branchName)
+		if err != nil {
+			ctx.ServerError("GetLatestPullRequestByHeadInfo", err)
+			return nil
+		}
+		if pr != nil {
+			pr.LoadIssue()
```
Comment (lunny) [inline routers/repo/branch.go]:
```go
if err := pr.LoadIssue(); err != nil {
  ctx.ServerError("pr.LoadIssue", err)
  return nil
}
```

**Comment by lunny** [comment]
Please see https://github.com/go-gitea/gitea/pull/6990/files#r294051752 and also CI pointed that. Otherwise LGTM.

**Comment by saitho** [comment]
Can we still get this into 1.9.0? Would help us a lot cleaning merged or closed branches. :)

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
I think we can move this to v1.9 since it finished before our freeze time but haven't assign milestone.

**Pull request merged**